Changes for Home assistant support

mretallack opened the pull request:
# [HomeAssistant] - Voice Control Integration

## Status: Ready for Merge

This PR isHome Assistant integration with Dicio. It is **now ready for production** but requires review, it has been been tested for the past 2 months.

**Note:** This implementation was developed with assistance from AWS Kiro AI to accelerate the initial development and explore the feasibility of Home Assistant integration.

## Overview

Adds a new skill that allows users to control and query Home Assistant entities using voice commands. Users can turn devices on/off, check status, and query person locations through natural language.

## Features Implemented

### Voice Commands

**Entity Control:**
- Turn entities on/off/toggle: `"Turn living room light on"`, `"Switch bedroom lamp off"`
- Check entity status: `"Get status of garage door"`, `"Check front door"`
- Person location: `"Where is the person Mark"`, `"What is Sarah location"`

### Supported Entity Types
- Lights, switches, covers, locks, fans, media players
- Person entities for location tracking
- Any entity that supports standard Home Assistant services

### Configuration
- Settings UI for Base URL and Access Token
- Entity mapping system (friendly name → entity ID)
- Secure token storage using DataStore

## Technical Implementation

### Files Added/Modified

**Core Implementation:**
- `app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/`
  - `HomeAssistantInfo.kt` - Skill registration
  - `HomeAssistantSkill.kt` - Main skill logic
  - `HomeAssistantOutput.kt` - Output handling
  - `HomeAssistantApi.kt` - REST API client
  - `HomeAssistantSettings.kt` - Settings UI

**Sentence Definitions:**
- `app/src/main/sentences/skill_definitions.yml` - Added sentence IDs
- `app/src/main/sentences/en/home_assistant.yml` - English patterns

**Proto/DataStore:**
- `app/src/main/proto/skill_settings_home_assistant.proto` - Settings schema

**Tests:**
- `app/src/test/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantSkillTest.kt` - 19 comprehensive tests

**Documentation:**
- `SETUP.md` - Complete setup guide for users

**Resources:**
- String resources for UI and speech output
- Icon for the skill

### Architecture

- Uses Home Assistant REST API for all operations
- Implements entity mapping for flexible friendly names
- Handles multiple entity domains (light, switch, cover, lock, person, etc.)
- Provides detailed error messages for troubleshooting

## Testing

✅ All 19 unit tests passing:
- Entity control (on/off/toggle) - 8 tests
- Status queries - 7 tests  
- Person location - 4 tests

Tests cover:
- All sentence pattern variations
- With/without "the" article
- Different command verbs (turn, switch, get, check, etc.)
- Edge cases and pattern conflicts

## Known Issues / TODO

### Before Merge:
- [x] **Error handling** - More robust network error handling
- [x] **UI/UX review** - Settings screen design
- [x] **Integration tests** - Test against real Home Assistant instance
- [x] **Code review** - General code quality and style
- [x] **Performance** - Test with large numbers of entities

### Pattern Conflicts:
- Avoided conflict with navigation skill by using `"where is the person <name>"` instead of `"where is <name>"`

### Limitations:
- Only supports basic on/off/toggle/status operations
- No support for setting specific values (brightness, temperature, etc.)
- No support for scenes or automations
- Person location returns raw state (not formatted nicely)

## Future Enhancements

Potential improvements for future iterations:
- Support for setting brightness, color, temperature
- Scene activation
- Automation triggering
- Better person location formatting (zone names)
- Support for climate entities (thermostats)
- Multi-entity commands ("Turn off all lights")
- Confirmation prompts for sensitive actions

## How to Test

1. Set up a Home Assistant instance (or use existing)
2. Build and install the app
3. Go to Settings > Skills > Home Assistant
4. Configure Base URL and Access Token
5. Add entity mappings
6. Try voice commands (see SETUP.md for examples)

## Screenshots

_TODO: Add screenshots of settings UI and example interactions_

## Related Issues

_Link to any related issues or discussions_

## Checklist

- [x] Code compiles without errors
- [x] All tests pass
- [x] Documentation added (SETUP.md)
- [x] Tested on physical device
- [x] Security review completed
- [x] Ready for merge

---


Feedback and suggestions are welcome!

mretallack commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantSkill.kt:
```diff
@@ -0,0 +1,292 @@
+    private fun generateNumberVariations(input: String): List<String> {
+        // Map number words to their digit and homophone variations
+        val numberMappings = mapOf(
+            "one" to listOf("1", "won"),
```
This is language dependent, but not sure if other languages have the same issue

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantApi.kt:
```diff
@@ -0,0 +1,63 @@
+object HomeAssistantApi {
+    @Throws(IOException::class)
+    suspend fun getAllStates(baseUrl: String, token: String): JSONArray {
+        val connection = URL("$baseUrl/api/states").openConnection() as HttpURLConnection
```
Please use OkHttp3 instead of Java raw connections. There are already utils to make json requests in the repo iirc.

mretallack replied:
Rewritten to use OkHttp3 with OkHttpClient, Request.Builder, and a shared executeRequest helper. Removed all HttpURLConnection usage.

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantInfo.kt:
```diff
@@ -0,0 +1,497 @@
+                    kotlinx.coroutines.GlobalScope.launch(Dispatchers.IO) {
+                        try {
+                            dataStore.updateData {
+                                android.util.Log.d("HomeAssistant", "DataStore update started")
```
Why do you use `android.util.Log` instead of importing `Log`? Also, create a `const val TAG = "HomeAssistantInfo"`

mretallack replied:
Added import android.util.Log, replaced all android.util.Log FQN calls with Log.d(TAG, ...) / Log.e(TAG, ...), added private const val TAG = "HomeAssistantInfo". Also replaced FQN kotlinx.coroutines.GlobalScope with a proper import.

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantInfo.kt:
```diff
@@ -0,0 +1,497 @@
+                                putExtra(Intent.EXTRA_TITLE, "home_assistant_config.yaml")
+                            }
+                            // Note: In a real implementation, you'd need to handle the file creation
+                            // through an activity result launcher
```
What? 😅 
I think this does not do anything

mretallack replied:
Removed the dead comment and the entire unused onExport lambda (which created an Intent that was never launched).

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantInfo.kt:
```diff
@@ -0,0 +1,497 @@
+}
+
+@Composable
+fun EntityMappingsEditor(
```
Put this in a separate file

mretallack replied:
Extracted EntityMappingsEditor, EntityMappingDialog, and EntityPickerDialog into a new EntityMappingsEditor.kt file.

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantInfo.kt:
```diff
@@ -0,0 +1,497 @@
+    baseUrl: String,
+    accessToken: String,
+    onMappingsChange: (List<EntityMapping>) -> Unit,
+    onExport: (String, String, List<EntityMapping>) -> Unit,
```
Unused param

mretallack replied:
Removed onExport parameter from EntityMappingsEditor signature and call site. Also removed the now-unused Intent import.

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantInfo.kt:
```diff
@@ -0,0 +1,497 @@
+                }
+            )
+
+            EntityMappingsEditor(
```
Is this an editor for a file that usually gets exported by home assistant? If yes, then it makes sense to keep a separate import/export. If not, then please remove this, as it should rather be an application-wide import/export.

mretallack replied:
I will remove it, It was so during development I could restore etc.. without having to recreate all the settings, an app wide backup/restore would be better (will add that to my TODO)

Stypox commented on app/src/main/sentences/en/home_assistant.yml:
```diff
@@ -0,0 +1,26 @@
+  - (turn|switch) (the )?.entity_name. toggle
+
+select_source:
+  - (turn|switch|set|tune|change) (the )?.entity_name. to .source_name.
```
LLMs get this wrong, but `()` is not needed around | groups. this is not a regex (also the others).
```suggestion
  - turn|switch|set|tune|change the? .entity_name. to .source_name.
```

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantSkill.kt:
```diff
@@ -0,0 +1,292 @@
+                    handleSelectSource(settings, mapping, sourceName)
+                }
+            }
+        } catch (e: FileNotFoundException) {
+            HomeAssistantOutput.EntityNotFound("unknown")
+        } catch (e: Exception) {
+            if (e.message?.contains("401") == true || e.message?.contains("403") == true) {
+                HomeAssistantOutput.AuthFailed()
+            } else {
+                HomeAssistantOutput.ConnectionFailed()
+            }
+        }
```
Let the exception fall through if you can't recognize it, instead of doing a catch-all. The rest of the Dicio infrastructure will show the error nicely.

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantSkill.kt:
```diff
@@ -0,0 +1,292 @@
+        )
+    }
+
+    private fun generateNumberVariations(input: String): List<String> {
```
Please use `dicio-numbers` instead. It won't handle homophone variations though, that's something that could be improved on `dicio-numbers`'s side.

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantSkill.kt:
```diff
@@ -0,0 +1,292 @@
+    private suspend fun handleSetState(
+        settings: SkillSettingsHomeAssistant,
+        mapping: EntityMapping,
+        action: String
```
Pass `ParsedAction` here, and call `parseAction` when calling this function in `HomeAssistant.SelectSource`

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantSkill.kt:
```diff
@@ -0,0 +1,292 @@
+        return variations.distinct()
+    }
+
+    private fun findBestSourceMatch(requested: String, available: List<String>): String? {
```
There is already a lot of code for computing string similarity in utils/, do not reinvent it

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantOutput.kt:
```diff
@@ -0,0 +1,278 @@
+                    modifier = Modifier.fillMaxWidth()
+                )
+                Text(
+                    text = "\"$requestedSource\"",
```
use a `R.string` with `"%1$s"`, so the "" are translatable

Stypox commented on app/src/main/kotlin/org/stypox/dicio/skills/homeassistant/HomeAssistantYamlUtils.kt:
```diff
@@ -0,0 +1,65 @@
+import java.io.InputStream
+import java.io.OutputStream
+
+object HomeAssistantYamlUtils {
```
Please use Kotlin serialization instead

mretallack replied:
file was deleted entirely as part of removing the import/export feature

Stypox commented on app/src/main/sentences/skill_definitions.yml:
```diff
@@ -134,3 +134,34 @@ skills:
             type: string
+
+  - id: home_assistant
+    specificity: high
```
Make it `medium`, otherwise "turn the torch on" of the torch skill would conflict with `set_state_on`

mretallack replied:
Changed home_assistant specificity from high to medium, has made it less accurate, but the other skills are are then not effected.

Stypox commented on app/src/test/kotlin/org/stypox/dicio/skills/homeassistant/TestSkillContext.kt:
```diff
@@ -0,0 +1,19 @@
+import org.dicio.skill.standard.util.MatchHelper
+import java.util.Locale
+
+class TestSkillContext(input: String) : SkillContext {
```
I think there is already a mock skill context somewhere (?)

mretallack replied:
Investigated — MockSkillContext exists but throws on all properties. TestSkillContext provides a working MatchHelper with input text, which is needed for sentence scoring tests. Kept it and added a KDoc comment explaining why both exist.

Stypox commented on docs/skills/home-assistant.md:
```diff
@@ -0,0 +1,258 @@
+# Dicio Setup Guide
```
You can add this page to https://github.com/DicioTeam/docs instead, but please do not entirely write it with LLMs

mretallack replied:
Remove and added seperate MR, still work in progress, https://github.com/DicioTeam/docs/pull/2

Stypox commented on .gitignore:
```diff
@@ -15,3 +15,6 @@
 fastlane_config.txt
+
+# Kiro specs (local only)
+.kiro/
```
What is kiro?

mretallack replied:
I will remove the line, Its the editor I am using.

Stypox replied:
I guess you can put `.kiro/` in your `.git/info/exclude`

mretallack replied:
Yes, I have done that now.

mretallack commented:
@Stypox  Thanks for the review, I have done the actions, Hopefully this is better, I have also tested on my phone.

mretallack committed:
Feature: add Home Assistant skill

Add a Home Assistant skill that allows voice control of HA entities.
Supports querying entity state, turning entities on/off, and selecting
media sources on media players.

- HomeAssistantSkill with fuzzy entity name matching
- Entity mapping configuration via settings UI
- Media source selection with number variation handling
- Help command listing available voice commands
- Comprehensive test coverage (fuzzy matching, number variations,
  source selection integration)

Co-authored-by: Mark <mark@retallack.org.uk>

mretallack commented:
@Stypox I gave rebased this branch to clean up the commits, hopefully the history looks a bit better
